Minor fixes to util/sss_krb5

alexey-tikhonov opened the pull request:
Main purpose of those patches is to resolve: https://pagure.io/SSSD/sssd/issue/4081

It also fixes one coverity issue:
```
Error: DEADCODE (CWE-561):
sssd-2.2.3/src/util/sss_krb5.c:375: cond_const: Condition "(kt_err = krb5_kt_next_entry(ctx, keytab, &entry, &cursor)) == 0", taking true branch. Now the value of "kt_err" is equal to 0.
sssd-2.2.3/src/util/sss_krb5.c:406: const: At condition "kt_err != 0", the value of "kt_err" must be equal to 0.
sssd-2.2.3/src/util/sss_krb5.c:406: dead_error_condition: The condition "kt_err != 0" cannot be true.
sssd-2.2.3/src/util/sss_krb5.c:406: dead_error_line: Execution cannot reach the expression "kt_err != -1765328202L" inside this statement: "if (kt_err != 0 && kt_err !...".
#  404|   
#  405|       /* check if we got any errors from krb5_kt_next_entry */
#  406|->     if (kt_err != 0 && kt_err != KRB5_KT_END) {
#  407|           DEBUG(SSSDBG_CRIT_FAILURE, "Error while reading keytab.\n");
#  408|           goto done;
```

sumit-bose commented on src/util/sss_krb5.c:
```diff
@@ -245,23 +241,20 @@ errno_t select_principal_from_keytab(TALLOC_CTX *mem_ctx,
-        DEBUG(SSSDBG_MINOR_FAILURE, "No suitable principal found in keytab\n");
-        ret = ENOENT;
+        error_message = "No suitable principal found in keytab";
+        ret = EINVAL;
```
Is there a reason you prefer EINVAL here? Currently the actual value of the error code is checked nowhere os it does not matter, but I still think ENOENT fits better.

alexey-tikhonov replied:
I understand initial intention, but `ENOENT == “No such file or directory”`  --  I think this is misleading as keytab file was found (and btw this error code was the root cause of initial bug report)

EINVAL == “Invalid argument”  --  might be not the best choice either, but I was unable to find anything more appropriate.

Would it be better to extend `utils_error.h:sssd_errors`?

sumit-bose replied:
Yes, a dedicated error would be good.

alexey-tikhonov replied:
Done.

sumit-bose commented on src/util/sss_krb5.c:
```diff
@@ -126,10 +126,12 @@ errno_t select_principal_from_keytab(TALLOC_CTX *mem_ctx,
               "Failed to read keytab [%s]: %s\n",
                KEYTAB_CLEAN_NAME,
-               sss_krb5_get_error_message(krb_ctx, kerr));
+               krb5_err_msg);
```
You do not handle the case where krb5_err_msg is NULL here as you did at a different place.

Maybe this can be made a macro since the pattern occurs at multiple places? But I'm not sure if it would work with complex format strings.

alexey-tikhonov replied:
> You do not handle the case where krb5_err_msg is NULL here as you did at a different place.

Thank you. Fixed.

sumit-bose commented on src/util/sss_krb5.c:
```diff
@@ -367,7 +360,10 @@ krb5_error_code find_principal_in_keytab(krb5_context ctx,
-        DEBUG(SSSDBG_CRIT_FAILURE, "krb5_kt_start_seq_get failed.\n");
+        const char *krb5_err_msg = sss_krb5_get_error_message(ctx, kerr);
+        DEBUG(SSSDBG_CRIT_FAILURE, "krb5_kt_start_seq_get failed: %s\n",
+              (krb5_err_msg ? krb5_err_msg : ""));
```
Maybe it would be better to print something like "- no error message available -" instead of ""?

alexey-tikhonov replied:
Fixed.

sumit-bose commented on src/util/sss_krb5.c:
```diff
@@ -416,9 +416,11 @@ krb5_error_code find_principal_in_keytab(krb5_context ctx,
-    if (kerr_d != 0) {
-        DEBUG(SSSDBG_CRIT_FAILURE, "Failed to free keytab entry.\n");
+    if (principal_found) {
+        kerr_d = sss_krb5_free_keytab_entry_contents(ctx, &entry);
```
I think strictly speaking this is not a double free since if the principal is not found entry should be memset to 0 and and the krb5_free_* calls should be able to handle that.

frozencemetery replied:
It might be worth moving the first `memset(&entry, 0, ...)` to the top of the function for clarity.

alexey-tikhonov replied:
> I think strictly speaking this is not a double free since if the principal is not found entry should be memset to 0 and and the krb5_free_* calls should be able to handle that.

Doc - https://web.mit.edu/kerberos/krb5-current/doc/appdev/refs/api/krb5_free_keytab_entry_contents.html - doesn't promise this.

From my point of view, it is easier to be on a safe side.

frozencemetery replied:
krb5 functions that free data all handle NULL, to match the behavior of `free()`.

alexey-tikhonov replied:
> krb5 functions that free data all handle NULL, to match the behavior of `free()`.

I am pretty sure they do. This is sane approach.
But is it stated officially in the docs somewhere?

alexey-tikhonov replied:
> It might be worth moving the first `memset(&entry, 0, ...)` to the top of the function for clarity.

Well, this is slight pessimization but really doesn't matter much. I do not have strong preference here... Ok, moved.

frozencemetery replied:
Any project that *doesn't* define free(NULL) is dangerous and needs to be fixed.  No, it's not explicitly stated because we're reasonable people.  There's only one libkrb5.

alexey-tikhonov replied:
If it is not in the API docs then it is implementation specific and thus can't be relied upon.
This is the only reasonable approach of any user of 3rd party lib.

sumit-bose commented on src/util/sss_krb5.c:
```diff
@@ -395,16 +395,16 @@ krb5_error_code find_principal_in_keytab(krb5_context ctx,
-    /* check if we got any errors from krb5_kt_next_entry */
-    if (kt_err != 0 && kt_err != KRB5_KT_END) {
-        DEBUG(SSSDBG_CRIT_FAILURE, "Error while reading keytab.\n");
+        /* (!principal_found) => (kt_err != 0) */
```
Can you make the comment a bit more explicit? At the first glance it looked a bit like commented out code.

alexey-tikhonov replied:
Fixed.

alexey-tikhonov commented:
@sumit-bose, @frozencemetery, thank you for a review.
I think I addressed all found issues.

alexey-tikhonov commented:
> I'm kind of confused that you think that, since we pretty clearly disagree: you claim that you've found a double-free, and I'm telling you that free(NULL) is fine and normal.

 `free(NULL)` is fine and normal because man page says so explicitly.

`krb5_free_keytab_entry_contents(structure_that_was_memset(0))` is not until docs explicitly say other way.

Lets put it other way: do you have any reasonable objection against patch?
I can edit commit message replacing "fixed error" with "put sanity guard" if you want.

frozencemetery commented:
Yes, of course I object!  You're claiming a double free; there's no double free.  You haven't found a security bug; you've found nothing at all.

alexey-tikhonov commented:
> Yes, of course I object! You're claiming a double free; there's no double free.

Ok, I see your point.

I have rewritten this patch entirely (to avoid treating failure of krb5_kt_end_seq_get() as an fatal error) and removed "double-free claim" from commit message.

Still I am bold in my statement that robust code should not rely on undocumented features (i.e. krb5_free_keytab_entry_contents(null_struct)) and this is still reflected in this patchset.

alexey-tikhonov commented:
Hi @sumit-bose, am I expected to fix anything else to make next review round possible?

sumit-bose commented:
Hi,

sorry for the delay. The code looks good and Coverity didn't find any new issues and said the deadcode issue is solved.

Also the debug messages are looking better now. But after checking the original ticket which asked for more specific debug messages I wonder if they  can be improved even further by
 - adding a sss_log messages with the final error message to find_principal_in_keytab() as well to make the reason why a keytab cannot be read (not exists, empty or other reason) more clear
 - if the default keytab is used to not just print 'default' in the debug/log messages in select_principal_from_keytab() but the real path of the default keytab.

Please let me know what you think?

bye,
Sumit

alexey-tikhonov commented:
Hi @sumit-bose,

please take a look at another attempt.

To address your second remark I have introduced `sss_printable_keytab_name()` helper.
If its impl is ok then I think we should make use of it instead of `KEYTAB_CLEAN_NAME` in ldap/krb5_child as well (but I think as another PR since it goes of scope)


Btw, during testing I have discovered that do-while loop in `select_principal_from_keytab()` is written in a such a way that it may try (primary=null)&(realm=null) at the last step. How do you think, is it an intent? I guess it is an error.

alexey-tikhonov commented:
I also think it is possible to stop trying different `primary`&`realm` after first fatal (i.e. not KRB5_KT_NOTFOUND) error returned by `find_principal_in_keytab()`.

But not sure if it is worth fixing.

alexey-tikhonov commented:
retest this please

alexey-tikhonov commented:
Hi,

> I only have a minor nitpick about ` static char buff[512];`. The `512` comes a bit out of the blue and one might ask why not 256 or 1024? I'd like to ask you to either add a comment that 512 is consider sufficient to store any kind of keytab description or use something like `sizeof("FILE:")+PATH_MAX` as size.

Thank you. I opted to add a comment (and rebased on current master).

In regards of PATH_MAX:
1) If I understand correctly, this value is merely max buffer size of some syscalls. It has nothing to do with actual maximum file path length (which can be literally arbitrary). Some details can be found [here](http://insanecoding.blogspot.com/2007/11/pathmax-simply-isnt.html) and [here](https://eklitzke.org/path-max-is-tricky). Thus IMO it doesn't actually guarantee that result will fit a buffer (but rather suggests a false assurance of this and so is a little bit misleading).
2) Additionally posix [doesn't guarantee](https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/limits.h.html) PATH_MAX to be constant/defined at all and (even if defined) different platforms use different headers for this definition, so it is tricky to use this properly.

alexey-tikhonov commented:
retest this please

sumit-bose commented:
Hi Alexey,

thanks for adding the comment. ACK.

bye,
Sumit

pbrezina commented:
* `master`
    * e778fa18a8d5f602c7d16087c42ec7861ffceb9f - util/sss_krb5: debug messages fixes
    * 716aebab5de67979f1afd5a14140042fb5e5567c - util/sss_krb5: fixed few memory handling issues
    * 8f275460a8831b2ab1fe3269d9a0ee3d88d3339f - util/sss_krb5: find_principal_in_keytab() was amended
    * 5086353ebad66d430aac40efff9778f0a0d2ce7c - util/sss_krb5.c: elimination of unreachable code
